### PR TITLE
Fix the MSAL build

### PR DESCRIPTION
**Pull request by heidijinxujia**
To fix issue #280 
- Add branch name to submodule
- Update the submodule to latest commit
- Update the build tool version and gradle version
- Replace the deprecated gradle commands

Filed two github issues to fix the broken unit tests, as fixing the unit tests is not the focus of this PR.

1. Fix the broken unit tests due to Common-core Logger Migration https://github.com/AzureAD/microsoft-authentication-library-for-android/issues/278
PR is out here #282 
2. Fix the broken unit tests due to Common-core Cache Migration https://github.com/AzureAD/microsoft-authentication-library-for-android/issues/279

### CODE REVIEW

**Comment by iambmelt** [comment]
@heidijinxujia Is there a reason we can't fix the `LoggerTests` in this PR?

**Comment by heidijinxujia** [comment]
@iambmelt The broken tests are caused by both common-core cache and common-core logger. I want to make them into separate PR. The aim of this PR is fixing the local build and travis build.

File: build.gradle [hunk 1]
```diff
@@ -3,6 +3,7 @@ buildscript {
 
     repositories {
         jcenter()
+        google()
```
Comment (heidijinxujia) [inline build.gradle]: @piqiums This is to fix this following problem during build.
```A problem occurred configuring root project 'msal-common'.
> Could not resolve all files for configuration ':classpath'.
   > Could not find com.android.tools.build:gradle:3.0.1.
     Searched in the following locations:
         https://jcenter.bintray.com/com/android/tools/build/gradle/3.0.1/gradle-3.0.1.pom
         https://jcenter.bintray.com/com/android/tools/build/gradle/3.0.1/gradle-3.0.1.jar
     Required by:
         project :```

**Comment by addev-ashish** [comment]
@heidijinxujia change the title of this bug and issue to fix Unit tests and build. we should never approve a PR that has an @Ignore on a unit test.

**Comment by heidijinxujia** [comment]
@addev-ashish I have made the fix for #278 in the PR #282.

**Comment by heidijinxujia** [comment]
@addev-ashish You mentioned there are several other concerns.

File: msal/src/request/java/com/microsoft/identity/client/BaseRequest.java [hunk 2]
```diff
@@ -204,7 +204,7 @@ AuthenticationResult postTokenRequest() throws MsalUiRequiredException, MsalServ
 
         final AccessTokenCacheItem accessTokenCacheItem;
         // Switch to toggle using common cache (or not)
-        boolean useCommonCache = true;
+        boolean useCommonCache = false;
```
Comment (addev-ashish) [inline msal/src/request/java/com/microsoft/identity/client/BaseRequest.java]: why have you turned this off?

File: testapps/automationapp/build.gradle [hunk 3]
```diff
@@ -46,6 +46,6 @@ android {
 }
 
 dependencies {
-    compile project (':msal')
-    compile "com.android.support:appcompat-v7:$rootProject.ext.supportLibraryVersion"
+    implementation project (':msal')
```
Comment (addev-ashish) [inline testapps/automationapp/build.gradle]: changing from compile to implementation threw several issues when i was building MSAL, why is this required?

File: msal/src/request/java/com/microsoft/identity/client/BaseRequest.java (hunk 2 above)
Comment (addev-ashish) [reply msal/src/request/java/com/microsoft/identity/client/BaseRequest.java]: This is by design. Sync'ed with Brian

File: testapps/automationapp/build.gradle (hunk 3 above)
Comment (addev-ashish) [reply testapps/automationapp/build.gradle]: I am not too hung up on this, if this builds I am good.

File: .gitmodules [hunk 4]
```diff
@@ -1,3 +1,4 @@
 [submodule "microsoft-authentication-library-common-for-android"]
 	path = common
 	url = https://github.com/AzureAD/microsoft-authentication-library-common-for-android.git
+	branch = dev
```
Comment (addev-ashish) [inline .gitmodules]: @iambmelt @heidijinxujia do we really want to do this?

File: msal/src/request/java/com/microsoft/identity/client/BaseRequest.java (hunk 2 above)
Comment (heidijinxujia) [reply msal/src/request/java/com/microsoft/identity/client/BaseRequest.java]: This change is from @iambmelt

File: testapps/automationapp/build.gradle (hunk 3 above)
Comment (heidijinxujia) [reply testapps/automationapp/build.gradle]: This change is coming from the compatibility requirements of common-core submodule. The common-core is using buildToolsVersion = "26.0.2" and gradle v3.0.1. That is why I upgrade the versions here. And compile is deprecated in 26 and it need to be replaced with implementation.

File: .gitmodules (hunk 4 above)
Comment (heidijinxujia) [reply .gitmodules]: If we leave the branch empty, by default, it will fetch the master branch of common-core which is an empty code base now. Without the branch config, the developer who first clones the MSAL repo will encounter the error on finding the :common project. 

Thus, we need to config the branch to dev in the .gitmodules.

File: msal/src/request/java/com/microsoft/identity/client/BaseRequest.java (hunk 2 above)
Comment (heidijinxujia) [reply msal/src/request/java/com/microsoft/identity/client/BaseRequest.java]: @iambmelt The unit tests for common-core cache is not fixed currently and here it only switched off the useCommonCache flag, right?

Comment (iambmelt) [reply msal/src/request/java/com/microsoft/identity/client/BaseRequest.java]: @heidijinxujia @addev-ashish I've switched off these tests because this branch uses the _old_ 'new' cache that Shane and I developed back in November/December of last year (before the new cache schema was authored). The branch `iambmelt/account-credential-schema` has the _new_ 'new' cache which has the schema updates and a different API.

File: msal/build.gradle [hunk 5]
```diff
@@ -103,20 +100,29 @@ task sourcesJar(type: Jar) {
...
+    androidTestImplementation "com.android.support.test:runner:$rootProject.ext.runnerVersion"
     // Set this dependency to use JUnit 4 rules
-    androidTestCompile "com.android.support.test:rules:$rootProject.ext.rulesVersion"
-    androidTestCompile "org.mockito:mockito-core:$rootProject.ext.mockitoCoreVersion"
-    androidTestCompile "com.google.dexmaker:dexmaker:$rootProject.ext.dexmakerMockitoVersion"
-    androidTestCompile "com.google.dexmaker:dexmaker-mockito:$rootProject.ext.dexmakerMockitoVersion"
-    compile project(':common')
+    androidTestImplementation "com.android.support.test:rules:$rootProject.ext.rulesVersion"
+    androidTestImplementation "org.mockito:mockito-android:$rootProject.ext.mockitoAndroidVersion"
+    api(project(":common")) {
```
Comment (iambmelt) [inline msal/build.gradle]: Looking into this further, it looks as if _all_ transitive dependencies can be excluded with:

```groovy
    implementation(project(':common')) {
        transitive = false
    }
```

This would be cleaner, let's do this instead.

File: .gitmodules (hunk 4 above)
Comment (iambmelt) [reply .gitmodules]: @heidijinxujia This isn't correct -- the submodule will fetch whichever commit the submodule pointed to when the last commit was made _containing_ the submodule.

The SHA doesn't need to be from a specific branch, it only needs to exist in the source tree. Usually you see this indicated by an `@ <sha>` in the filename

See below:
capture

All this to say, let's **not** explicitly point the submodule reference to a branch as it will make feature branches cumbersome and potentially even _more_ confusing

**Comment by iambmelt** [comment]
@heidijinxujia See above feedback re: submodule branch reference. Let's keep the default behavior for now and _not_ point it towards a branch without a specific commit 

/cc @shoatman Since this affects you too

File: .gitmodules (hunk 4 above)
Comment (heidijinxujia) [reply .gitmodules]: Sorry, I probably I have a different thought.

This is a more robust way to update the working tree of the submodules and get rid of fetching the submodule change in every PR. If we point the gitmodules to a branch, for example 'dev', we will not need to commit the commin@SHA everytime. Instead, git gives us a way to fetch the change automatically by calling
`git submodule init`
`git submodule update --remote`

Well, it quite depends the usage of the submodule. As the common repo is not released yet, I would like to leave the decision to you and Shane.

File: msal/build.gradle [hunk 6]
```diff
@@ -103,20 +100,21 @@ task sourcesJar(type: Jar) {
...
+    androidTestImplementation "com.android.support.test:runner:$rootProject.ext.runnerVersion"
     // Set this dependency to use JUnit 4 rules
-    androidTestCompile "com.android.support.test:rules:$rootProject.ext.rulesVersion"
-    androidTestCompile "org.mockito:mockito-core:$rootProject.ext.mockitoCoreVersion"
-    androidTestCompile "com.google.dexmaker:dexmaker:$rootProject.ext.dexmakerMockitoVersion"
-    androidTestCompile "com.google.dexmaker:dexmaker-mockito:$rootProject.ext.dexmakerMockitoVersion"
-    compile project(':common')
+    androidTestImplementation "com.android.support.test:rules:$rootProject.ext.rulesVersion"
+    androidTestImplementation "org.mockito:mockito-android:$rootProject.ext.mockitoAndroidVersion"
+    implementation(project(':common')) {
```
Comment (iambmelt) [inline msal/build.gradle]: ah! I see you've already taken care of this 😎

**Comment by heidijinxujia** [comment]
Thank you @iambmelt !

File: .gitmodules (hunk 4 above)
Comment (iambmelt) [reply .gitmodules]: My preference would be to _not_ specify a branch and to instead let it point to a commit SHA. My rationale is that this makes it easier for feature branches to point to other branches/commits while in developments and also that it reduces the likelihood of something changing the behavior or api of a submodule depedency 'beneath' this repo. Example: Someone modifies the API of class in `common` without updating calls to it from `MSAL` (or elsewhere).

@shoatman ^^ Take a look at the above and weigh-in

**Comment by iambmelt** [comment]
Let's get @shoatman to look at the submodule config -- I suspect he'll also want to use explicit commits instead of branch references but let's double check before we merge just in case

**Comment by heidijinxujia** [comment]
@shoatman Could you have a look at this PR when you avail? Thanks!
